# Thursday May 15, 2024

## Meeting summary

We're going to include F:B ratios to help explain some of the trends. The paper will be mostly descriptive and will have several small stories instead of having a big overall story. We will also place less emphasis on the Vmax dataset and moreso on the CAZyme genes.

The discussion will have the following sections

1. Differences between the vegetation communities
2. Differences between drought treatments
	1. Effects for coastal sage scrub
	2. Effects for grass
3. Temporal trends



The patterns we will describe are
* Higher F:B ratios at the beginning that then decrease with time, and this corresponds with peptidoglycan genes and chitin genes, and that this describes successional changes in the microbial communities (in the temporal trends section of the discussion). Bacteria have peptidoglycan cell walls, and high F:B ratios at the beginning and low peptidoglycan genes at the beginning means there are not enough bacteria to act as a substrate to select for peptidoglycan genes. Decreasing trend in chitin genes correspond with decreasing F:B ratios because there are less fungi (whose cell walls are made out of chitin) to select for chitin genes
* Temporal trend in F:B ratios also explain decreasing trend in cellulose genes with time, because there are less fungi
* Personally, I don't think that the metagenomic dataset can be linked to the litter chemistry dataset very strongly, but Ashish wants to link them anyway by focusing on timepoints 1 and 2 of polysaccharides with C-O stretching. For these 2 timepoints in shrub, the drought treatment has a decreasing trend for both polysaccharide genes and carbohydrate C-O stretching while the ambient treatment has an increasing trend for both. Same thing with carbo3 (carbo-ester) in shrub


Points I brought up
* Decrease in lignin genes due to drought caused by drought selecting against lignin genes
	- Outcome: Ashish shot this down, he believes very strongly that genes are selected by the availability of substrates, so the decrease in lignin genes is due to potential decreases in lignin
* Alkane functional group describing lignin
	- Outcome: Ashish shot this down. He says that the FTIR is too low resolution, and that the people at UC Berkeley who did the FTIR said that the signal for lignin is hard to detect in FTIR
	- Let me see if I can convince him with a couple of papers
	

Action items
* Brian: make figures in powerpoint by cropping them together
* Brian: send Ashish log10-transformed dataset exactly how Ashish formats his datasets
* Brian: ask Steve for a meeting the week of Monday June 17th, 2024
* Ashish: send enzyme Vmax figures and CAZyme figures as .png
* Both: make presentation for meeting with Steve